bee-ledger/src/workers/pruning/batch.rs

```diff
@@ -277,115 +292,223 @@ pub fn batch_prunable_unconfirmed_data<S: StorageBackend>(
             }
         }
 
-        Batch::<(MilestoneIndex, UnreferencedMessage), ()>::batch_delete(
-            storage,
-            batch,
-            &(prune_index, (*unconf_msg_id).into()),
-        )
-        .map_err(|e| Error::Storage(Box::new(e)))?;
+        byte_length += prune_unreferenced_message(storage, batch, prune_index, (*unconf_msg_id).into(), by_size)?;
 
         metrics.prunable_messages += 1;
     }
 
-    Ok(metrics)
+    Ok((byte_length, metrics))
 }
 
-pub fn prune_milestone_data<S: StorageBackend>(
+pub(crate) fn prune_milestone_data<S: StorageBackend>(
     storage: &S,
     batch: &mut S::Batch,
     prune_index: MilestoneIndex,
     should_prune_receipts: bool,
-) -> Result<MilestoneDataPruningMetrics, Error> {
+    by_size: bool,
+) -> Result<(ByteLength, MilestoneDataPruningMetrics), PruningError> {
+    let mut byte_length = 0usize;
     let mut metrics = MilestoneDataPruningMetrics::default();
 
-    prune_milestone(storage, batch, prune_index)?;
-
-    prune_output_diff(storage, batch, prune_index)?;
+    byte_length += prune_milestone(storage, batch, prune_index, by_size)?;
+    byte_length += prune_output_diff(storage, batch, prune_index, by_size)?;
 
     if should_prune_receipts {
-        metrics.receipts = prune_receipts(storage, batch, prune_index)?;
+        let (num_receipts, num_bytes) = prune_receipts(storage, batch, prune_index, by_size)?;
+
+        metrics.receipts = num_receipts;
+        byte_length += num_bytes;
     }
 
-    Ok(metrics)
+    Ok((byte_length, metrics))
 }
 
 fn prune_message_and_metadata<S: StorageBackend>(
     storage: &S,
     batch: &mut S::Batch,
     message_id: &MessageId,
-) -> Result<(), Error> {
-    Batch::<MessageId, Message>::batch_delete(storage, batch, message_id).map_err(|e| Error::Storage(Box::new(e)))?;
+    by_size: bool,
+) -> Result<ByteLength, PruningError> {
+    let mut byte_length = 0usize;
+
+    if by_size {
+        // Panic: we know the message is in the database.
```

Comment:
Given that there is no DB lock, is that the case?

Reply:
Pruning would have panicked already if it weren't there before this line is executed. And this pruning is the only "thing" that can remove stuff from the otherwise append-only database...this cannot panic here anymore.

Reply:
oh OK that makes a lot of sense. I'd extend this comment saying that pruning is the only thing able to remove messages to make it clearer